### PR TITLE
[mlir] Support DialectRegistry extension comparison

### PR DESCRIPTION
**Pull request by nikalra**
`PassManager::run` loads the dependent dialects for each pass into the current context prior to invoking the individual passes. If the dependent dialect is already loaded into the context, this should be a no-op. However, if there are extensions registered in the `DialectRegistry`, the dependent dialects are unconditionally registered into the context.

This poses a problem for dynamic pass pipelines, however, because they will likely be executing while the context is in an immutable state (because of the parent pass pipeline being run).

To solve this, we'll update the extension registration API on `DialectRegistry` to require a type ID for each extension that is registered. Then, instead of unconditionally registered dialects into a context if extensions are present, we'll check against the extension type IDs already present in the context's internal `DialectRegistry`. The context will only be marked as dirty if there are net-new extension types present in the `DialectRegistry` populated by `PassManager::getDependentDialects`.

Note: this PR removes the `addExtension` overload that utilizes `std::function` as the parameter. This is because `std::function` is copyable and potentially allocates memory for the contained function so we can't use the function pointer as the unique type ID for the extension.

Downstream changes required:
- Existing `DialectExtension` subclasses will need a type ID to be registered for each subclass. More details on how to register a type ID can be found here: https://github.com/llvm/llvm-project/blob/8b68e06731e0033ed3f8d6fe6292ae671611cfa1/mlir/include/mlir/Support/TypeID.h#L30
- Existing uses of the `std::function` overload of `addExtension` will need to be refactored into dedicated `DialectExtension` classes with associated type IDs. The attached `std::function` can either be inlined into or called directly from `DialectExtension::apply`.

### CODE REVIEW

**Review by joker-eph — CHANGES_REQUESTED** [review]
Whether threading is enabling or not should not affect the invariants of the pass manager, the assert is very intentional in its current form.

Seems like an issue with detecting that some loading is a no-op?

**Comment by nikalra** [comment]
> Whether threading is enabling or not should not affect the invariants of the pass manager, the assert is very intentional in its current form.
> 
> Seems like an issue with detecting that some loading is a no-op?

It seems like this should be a better approach by allowing us to check if extensions have already been registered into the context (instead of conservatively assuming they aren't). But, this change touches 90+ extra files, so if there's a way to do this more cleanly I'm more than happy to investigate that further!

**Comment by joker-eph** [comment]
Thanks for giving it a try! I feel that relying on a string provided by the user is not desirable. We need to revisit this with the same kind of mechanism we used elsewhere (based on TypeID / global symbol address) somehow.

File: mlir/include/mlir/IR/DialectRegistry.h
Comment (nikalra) [inline mlir/include/mlir/IR/DialectRegistry.h:234]: We'll have to remove the `std::function` overload of `addExtension` so we can use the underlying function pointer as the TypeID, but the lost functionality can be achieved with a `DialectExtension` subclass if needed.

File: mlir/include/mlir/IR/DialectRegistry.h
Comment (nikalra) [inline mlir/include/mlir/IR/DialectRegistry.h:261]: If the overall theme of this change looks good, I'd like to also convert this into a `DenseMap` so we always de-duplicate by key (instead of doing that implicitly in code paths that call `isSubset`).

It'll make `applyExtensions` slightly less efficient since we'll have to track changes in case additional extensions are registered when applying the existing extensions set, but I think the code clarity would probably be worth it.

**Review by nikalra — COMMENTED** [review]
Hopefully this is closer to what you had in mind!

File: mlir/lib/IR/Dialect.cpp
Comment (River707) [inline mlir/lib/IR/Dialect.cpp:299]: llvm::make_first_range

File: mlir/lib/IR/Dialect.cpp
Comment (River707) [inline mlir/lib/IR/Dialect.cpp:296]: You can drop the `llvm::` on all of these, they should already be re-exported into the mlir namespace.

File: mlir/lib/IR/Dialect.cpp
Comment (River707) [inline mlir/lib/IR/Dialect.cpp:300]: You could use llvm::set_union here.

File: mlir/include/mlir/IR/DialectRegistry.h
Comment (River707) [reply mlir/include/mlir/IR/DialectRegistry.h:261]: You can't use a DenseMap here, that will make iteration order non-deterministic, this could be a MapVector though (which still gives you random access via the begin() iterator).

**Review by River707 — COMMENTED** [review]
This looks a lot cleaner! Will this require any changes to existing in-tree usages? I do appreciate not needing to change ~every current usage.

**Comment by nikalra** [comment]
> This looks a lot cleaner! Will this require any changes to existing in-tree usages? I do appreciate not needing to change ~every current usage.

I think just a handful: we should get most of the lambdas for free since I don't think any of them have captures, but we'll have to add type ID macros to the DialectExtension subclasses.

File: mlir/include/mlir/IR/DialectRegistry.h
Comment (nikalra) [reply mlir/include/mlir/IR/DialectRegistry.h]: Ah, I didn't realize iteration order was important. Switched to MapVector!

File: mlir/lib/IR/Dialect.cpp
Comment (joker-eph) [inline mlir/lib/IR/Dialect.cpp:209]:
```suggestion
    for (const auto *ext : subset)
      applyExtension(*ext);
```

**Review by joker-eph — COMMENTED** [review]
Looks great! Thanks :)

**Comment by joker-eph** [comment]
Can you add in the PR description a note about the removed API and the advice on how to proceed for people affected (how to replace it).

**Comment by nikalra** [comment]
> Can you add in the PR description a note about the removed API and the advice on how to proceed for people affected (how to replace it).

Done!

**Comment by nikalra** [comment]
If this change looks good, would it be possible to merge it in? I don't have merge permissions for LLVM. Thanks!

**Review by joker-eph — APPROVED** [review]

**Comment by joker-eph** [comment]
(Waiting for CI to complete)

**Comment by joker-eph** [comment]
> Can you add in the PR description a note about the removed API and the advice on how to proceed for people affected (how to replace it)

Actually this seems missing? Or maybe I wasn't clear: downstream folks who will broken by this patch would appreciate in the description a step-by-step list of actions they need to take to upgrade their codebase.

**Comment by nikalra** [comment]
> > Can you add in the PR description a note about the removed API and the advice on how to proceed for people affected (how to replace it)
> 
> Actually this seems missing? Or maybe I wasn't clear: downstream folks who will broken by this patch would appreciate in the description a step-by-step list of actions they need to take to upgrade their codebase.

Added step-by-step instructions for the two modified use cases: existing DialectExtensions and the removed overload. Please let me know if I'm missing anything there or if there's anything that won't be clear to affected users!

**Comment by joker-eph** [comment]
Perfect, thanks!

**Pull request merged**
